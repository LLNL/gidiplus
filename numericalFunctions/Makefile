# <<BEGIN-copyright>>
# Copyright 2019, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: MIT
# <<END-copyright>>

PYTHON = python

SOURCE_DIRS = ptwC ptwX ptwXY nf_Legendre nf_specialFunctions nf_integration
SUB_DIRS = include lib $(SOURCE_DIRS) Doc _Src

.PHONY: default purgeIncludeAndLib includeAndLib $(SOURCE_DIRS) python utilities check checke checkPython checkePython \
	collectSources clean realclean doTarget

SMR_PATH = $(abspath ../statusMessageReporting)
SMR_INCLUDE_PATH = $(SMR_PATH)/include
SMR_LIB_PATH = $(SMR_PATH)/lib

MAKEFLAGS += SMR_PATH=$(SMR_PATH)

default: purgeIncludeAndLib $(SOURCE_DIRS)
	$(MAKE) utilities

purgeIncludeAndLib:
	cd include; $(MAKE) realclean
	cd lib; $(MAKE) realclean

includeAndLib:
	cd include; $(MAKE) NFUPDATING_PATH=$(NFUPDATING_PATH)
	cd lib; $(MAKE) NFUPDATING_PATH=$(NFUPDATING_PATH)

ptwC:
	cd ptwC; $(MAKE)
	$(MAKE) includeAndLib NFUPDATING_PATH=../ptwC

ptwX: ptwC
	cd ptwX; $(MAKE)
	$(MAKE) includeAndLib NFUPDATING_PATH=../ptwX

ptwXY: ptwX
	cd ptwXY; $(MAKE)
	$(MAKE) includeAndLib NFUPDATING_PATH=../ptwXY

nf_Legendre: ptwXY
	cd nf_Legendre; $(MAKE)
	$(MAKE) includeAndLib NFUPDATING_PATH=../nf_Legendre

nf_specialFunctions: nf_Legendre
	cd nf_specialFunctions; $(MAKE)
	$(MAKE) includeAndLib NFUPDATING_PATH=../nf_specialFunctions

nf_integration: nf_specialFunctions                         # The chained dependencies are needed for the -j option. Why, do not know but does not work otherwise.
	cd nf_integration; $(MAKE)
	$(MAKE) includeAndLib NFUPDATING_PATH=../nf_integration

python:
	$(PYTHON) setup.py --quiet build

utilities:
	$(MAKE) doTarget MODE=utilities SUB_DIRS="$(SOURCE_DIRS)"

check:
	$(MAKE) doTarget MODE=check SUB_DIRS="$(SOURCE_DIRS)"

checke:
	$(MAKE) check CHECKOPTIONS=-e

checkPython:
	$(MAKE) doTarget MODE=checkPython SUB_DIRS="$(SOURCE_DIRS)"

checkePython:
	$(MAKE) CHECKOPTIONS=-e checkPython

collectSources:
	cd _Src; $(MAKE) collectSources

clean:
	rm -rf build
	$(MAKE) doTarget MODE=clean

realclean:
	rm -rf build *.pyc __pycache__
	$(MAKE) doTarget MODE=realclean

doTarget:
	SAVED_PWD=`pwd`; \
	for directory in $(SUB_DIRS); do cd $$directory; $(MAKE) $(MODE); cd $$SAVED_PWD; done
