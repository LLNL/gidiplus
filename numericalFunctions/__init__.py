# <<BEGIN-copyright>>
# Copyright 2019, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: MIT
# <<END-copyright>>

from .lib import pointwiseXY_C
from .lib import Legendre
from .lib.pointwiseXY_C import pointwiseXY_C as pointwiseXY
from .lib import specialFunctions
from .lib import angularMomentumCoupling
from .lib import integration
from .lib import listOfDoubles_C
