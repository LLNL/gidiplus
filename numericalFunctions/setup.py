# <<BEGIN-copyright>>
# Copyright 2019, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: MIT
# <<END-copyright>>

import os, glob, shutil
from distutils.core import setup, Extension

extra_compile_args = [ ]
# Option need for some MACs
# extra_compile_args = [ '-Wno-error=unused-command-line-argument-hard-error-in-future' ]

statusMessageReportingRoot = '../statusMessageReporting'

statusMessageReporting_c = glob.glob( os.path.join( statusMessageReportingRoot, 'Src', '*.c' ) )
ptwC_c = glob.glob( os.path.join( 'ptwC', 'Src', '*.c' ) )
ptwX_c = glob.glob( os.path.join( 'ptwX', 'Src', '*.c' ) )
ptwX_Py_c = glob.glob( os.path.join( 'ptwX', 'Python', 'Src', '*.c' ) )
ptwXY_c = glob.glob( os.path.join( 'ptwXY', 'Src', '*.c' ) )
ptwXY_Py_c = glob.glob( os.path.join( 'ptwXY', 'Python', 'Src', '*.c' ) )
nf_Legendre_c = glob.glob( os.path.join( 'nf_Legendre', 'Src', '*.c' ) )
nf_Legendre_Py_c = glob.glob( os.path.join( 'nf_Legendre', 'Python', 'Src', '*.c' ) )
nf_specialFunctions_c = glob.glob( os.path.join( 'nf_specialFunctions', 'Src', 'nf_[egip]*.c' ) )
nf_specialFunctions_Py_c = glob.glob( os.path.join( 'nf_specialFunctions', 'Python', 'Src', 'nf_specialFunctions_C.c' ) )
nf_angularMomentumCoupling_c = glob.glob( os.path.join( 'nf_specialFunctions', 'Src', 'nf_angularMomentumCoupling.c' ) )
nf_angularMomentumCoupling_Py_c = glob.glob( os.path.join( 'nf_specialFunctions', 'Python', 'Src', 'nf_angularMomentumCoupling_C.c' ) )
nf_integration_c = glob.glob( os.path.join( 'nf_integration', 'Src', '*.c' ) )
nf_integration_Py_c = glob.glob( os.path.join( 'nf_integration', 'Python', 'Src', '*.c' ) )

statusMessageReporting_hDir = os.path.join( statusMessageReportingRoot, 'Src' )
ptwC_hDir = os.path.join( 'ptwC', 'Src' )
ptwX_hDir = os.path.join( 'ptwX', 'Src' )
ptwXY_hDir = os.path.join( 'ptwXY', 'Src' )
ptwXY_Py_hDir =  os.path.join( 'ptwXY', 'Python', 'Src' )
nf_Legendre_hDir = os.path.join( 'nf_Legendre', 'Src' )
nf_specialFunctions_hDir = os.path.join( 'nf_specialFunctions', 'Src' )
nf_specialFunctions_Py_hDir = os.path.join( 'nf_specialFunctions', 'Python', 'Src' )
nf_integration_hDir = os.path.join( 'nf_integration', 'Src' )

#
# Stuff to build listOfDoubles.so.
#
libs = glob.glob( os.path.join( 'build', 'lib*', 'listOfDoubles_C*' ) )
for lib in libs : os.remove( lib )

listOfDoubles_C = Extension( 'listOfDoubles_C',
    extra_compile_args = extra_compile_args,
    sources = statusMessageReporting_c + ptwC_c + ptwX_c + ptwX_Py_c,
    include_dirs = [ statusMessageReporting_hDir, ptwC_hDir, ptwX_hDir ] )

setup( name = 'listOfDoubles_C', 
    version = '1.0',
    description = 'This module contains the listOfDoubles_C class and support routines.',
    ext_modules = [ listOfDoubles_C ] )

libs = glob.glob( os.path.join( 'build', 'lib*', 'listOfDoubles_C*' ) )
if( len( libs ) > 0 ) : shutil.copy( libs[0], 'lib' )

#
# Stuff to build pointwiseXY_C.so.
#
libs = glob.glob( os.path.join( 'build', 'lib*', 'pointwiseXY_C*' ) )
for lib in libs : os.remove( lib )

pointwiseXY_C = Extension( 'pointwiseXY_C',
    extra_compile_args = extra_compile_args,
    sources = statusMessageReporting_c + ptwC_c + ptwX_c + nf_Legendre_c + nf_integration_c + ptwXY_c + ptwXY_Py_c,
    include_dirs = [ statusMessageReporting_hDir, ptwC_hDir, ptwX_hDir, nf_Legendre_hDir, nf_integration_hDir, 
            ptwXY_hDir, ptwXY_Py_hDir ] )

setup( name = 'pointwiseXY_C', 
    version = '1.0',
    description = 'This module contains the pointwiseXY_C class and support routines.',
    ext_modules = [ pointwiseXY_C ] )

libs = glob.glob( os.path.join( 'build', 'lib*', 'pointwiseXY_C*' ) )
if( len( libs ) > 0 ) : shutil.copy( libs[0], 'lib' )

#
# Stuff to build Legendre.so.
#
libs = glob.glob( os.path.join( 'build', 'lib*', 'Legendre*' ) )
for lib in libs : os.remove( lib )

nf_Legendre_C = Extension( 'Legendre',
    extra_compile_args = extra_compile_args,
    sources = statusMessageReporting_c + ptwC_c + ptwX_c + ptwXY_c + nf_integration_c + nf_Legendre_c + nf_Legendre_Py_c,
    include_dirs = [ statusMessageReporting_hDir, ptwC_hDir, ptwX_hDir, ptwXY_hDir, ptwXY_Py_hDir, 
            nf_integration_hDir, nf_Legendre_hDir ] )

setup( name = 'Legendre', 
    version = '1.0',
    description = 'This module contains the Legendre class and support routines.',
    ext_modules = [ nf_Legendre_C ] )

libs = glob.glob( os.path.join( 'build', 'lib*', 'Legendre*' ) )
if( len( libs ) > 0 ) : shutil.copy( libs[0], 'lib' )

#
# Stuff to build specialFunctions.so
#
libs = glob.glob( os.path.join( 'build', 'lib*', 'specialFunctions*' ) )
for lib in libs : os.remove( lib )

specialFunctions = Extension( 'specialFunctions',
    extra_compile_args = extra_compile_args,
    sources = statusMessageReporting_c + ptwC_c + nf_specialFunctions_c + nf_specialFunctions_Py_c,
    include_dirs = [ statusMessageReporting_hDir, ptwC_hDir, nf_specialFunctions_hDir, nf_specialFunctions_Py_hDir ] )

setup( name = 'specialFunctions', 
    version = '1.0',
    description = 'This module contains some special math functions not in the python math module.',
    ext_modules = [ specialFunctions ] )

libs = glob.glob( os.path.join( 'build', 'lib*', 'specialFunctions*' ) )
if( len( libs ) > 0 ) : shutil.copy( libs[0], 'lib' )

#
# Stuff to build angularMomentumCoupling.so
#
libs = glob.glob( os.path.join( 'build', 'lib*', 'angularMomentumCoupling*' ) )
for lib in libs : os.remove( lib )

angularMomentumCoupling = Extension( 'angularMomentumCoupling',
    extra_compile_args = extra_compile_args,
    sources = statusMessageReporting_c + ptwC_c + nf_angularMomentumCoupling_c + nf_angularMomentumCoupling_Py_c ,
    include_dirs = [ statusMessageReporting_hDir, ptwC_hDir, nf_specialFunctions_hDir, nf_specialFunctions_Py_hDir ] )

setup( name = 'angularMomentumCoupling', 
    version = '1.0',
    description = 'This module contains some physics angular momentum coupling functions not in the python math module.',
    ext_modules = [ angularMomentumCoupling ] )

libs = glob.glob( os.path.join( 'build', 'lib*', 'angularMomentumCoupling*' ) )
if( len( libs ) > 0 ) : shutil.copy( libs[0], 'lib' )

#
# Stuff to build integration.so
#
libs = glob.glob( os.path.join( 'build', 'lib*', 'integration*' ) )
for lib in libs : os.remove( lib )

integration = Extension( 'integration',
    extra_compile_args = extra_compile_args,
    sources = statusMessageReporting_c + ptwC_c + ptwX_c + ptwXY_c + nf_Legendre_c + nf_integration_c + nf_integration_Py_c,
    include_dirs = [ statusMessageReporting_hDir, ptwC_hDir, ptwX_hDir, ptwXY_hDir, nf_Legendre_hDir, nf_integration_hDir ] )

setup( name = 'integration', 
    version = '1.0',
    description = 'This module contains functions for integrating a function representing an integrand.',
    ext_modules = [ integration ] )

libs = glob.glob( os.path.join( 'build', 'lib*', 'integration*' ) )
if( len( libs ) > 0 ) : shutil.copy( libs[0], 'lib' )
