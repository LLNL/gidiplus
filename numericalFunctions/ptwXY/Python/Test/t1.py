#! /usr/bin/env python

# <<BEGIN-copyright>>
# Copyright 2019, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: MIT
# <<END-copyright>>

"""
Some simple stuff that should repeat.
"""

from numericalFunctions import pointwiseXY_C

f = pointwiseXY_C.pointwiseXY_C( initialSize = 40, overflowSize = 10 )
print( 'len( f ) =', len( f ) )
print( 'f.allocatedSize( ) =', f.allocatedSize( ) )
print( 'f.overflowAllocatedSize( ) =', f.overflowAllocatedSize( ) )
print( 'f.overflowLength( ) =', f.overflowLength( ) )

data = [ [ 1, 1. ], [ 3., 3.14 ], [ 4.5, -5 ], [ 6.6, 2 ], [ 7.5, 4], [ 8.6, 2 ], [ 11.3, -5 ], [ 12.6, 3 ] ]
f.setData( data )
print( 'len( f ) =', len( f ) )

print( 'f[2] = ', f[2] )

print( 'f.evaluate( 2. ) =', f.evaluate( 2. ) )
print( 'f.evaluate( 5.5 ) =', f.evaluate( 5.5 ) )
print( 'f.evaluate( 7. ) =', f.evaluate( 7. ) )

print( 'f.domainMin( ) = ', f.domainMin( ) )
print( 'f.domainMax( ) = ', f.domainMax( ) )

print( )
print( "printing each element of f" )
for i in f : print( "  ", i )

print( f.toString.__doc__ )
print( )
print( "printing f" )
print( f )

print( 'printing f.toString( pairsPerLine = 3 )' )
print( f.toString( pairsPerLine = 3 ) )
print( """printing f.toString( format = " %6.3f %10.3e%%", pairsPerLine = 2, pairSeparator = ',' )""" )
print( f.toString( format = " %6.3f %10.3e%%", pairsPerLine = 2, pairSeparator = ',' ) )
print( """printing f.toString( format = " %6.3e %10.3E", pairSeparator = ',' )""" )
print( f.toString( format = " %6.3e %10.3E", pairSeparator = ',' ) )
