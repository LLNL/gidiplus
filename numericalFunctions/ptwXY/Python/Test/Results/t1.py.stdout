len( f ) = 0
f.allocatedSize( ) = 40
f.overflowAllocatedSize( ) = 10
f.overflowLength( ) = 0
len( f ) = 8
f[2] =  [4.5, -5.0]
f.getValue( 2. ) = 2.07
f.getValue( 5.5 ) = -1.66666666667
f.getValue( 7. ) = 2.88888888889
f.xMin( ) =  1.0
f.xMax( ) =  12.6

printing each element of f
   [1.0, 1.0]
   [3.0, 3.1400000000000001]
   [4.5, -5.0]
   [6.5999999999999996, 2.0]
   [7.5, 4.0]
   [8.5999999999999996, 2.0]
   [11.300000000000001, -5.0]
   [12.6, 3.0]
Returns a string representation of self. This method has three keyword parameters:
pairsPerLine, format and pairSeparator which are defined as,
    pairsPerLine    the number of pairs to put on each line
    format          a valid format to convert an (x,y) pair (i.e., two floats) into a string (e.g. format = ' %.3f %12.5e')
    pairSeparator   a string to put between every pair (e.g, to put a comma to separate pairs use pairSeparator = ',')

printing f
   1.00000000e+00   1.00000000e+00
   3.00000000e+00   3.14000000e+00
   4.50000000e+00  -5.00000000e+00
   6.60000000e+00   2.00000000e+00
   7.50000000e+00   4.00000000e+00
   8.60000000e+00   2.00000000e+00
   1.13000000e+01  -5.00000000e+00
   1.26000000e+01   3.00000000e+00

printing f.toString( pairsPerLine = 3 )
   1.00000000e+00   1.00000000e+00   3.00000000e+00   3.14000000e+00   4.50000000e+00  -5.00000000e+00
   6.60000000e+00   2.00000000e+00   7.50000000e+00   4.00000000e+00   8.60000000e+00   2.00000000e+00
   1.13000000e+01  -5.00000000e+00   1.26000000e+01   3.00000000e+00

printing f.toString( format = " %6.3f %10.3e%%", pairsPerLine = 2, pairSeparator = ',' )
  1.000  1.000e+00%,  3.000  3.140e+00%,
  4.500 -5.000e+00%,  6.600  2.000e+00%,
  7.500  4.000e+00%,  8.600  2.000e+00%,
 11.300 -5.000e+00%, 12.600  3.000e+00%

printing f.toString( format = " %6.3e %10.3E", pairSeparator = ',' )
 1.000e+00  1.000E+00,
 3.000e+00  3.140E+00,
 4.500e+00 -5.000E+00,
 6.600e+00  2.000E+00,
 7.500e+00  4.000E+00,
 8.600e+00  2.000E+00,
 1.130e+01 -5.000E+00,
 1.260e+01  3.000E+00

